pkg/backend/net/grpc.go

```diff
@@ -0,0 +1,97 @@
+package net
+
+import (
+	"context"
+	"github.com/bookingcom/carbonapi/pkg/types"
+	capi_v2_grpc "github.com/go-graphite/protocol/carbonapi_v2_grpc"
+	"github.com/go-graphite/protocol/carbonapi_v2_pb"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"io"
+	"time"
+)
+
+// GrpcBackend represents a host that accepts requests for metrics over gRPC.
+type GrpcBackend struct {
+	*Backend
+	carbonV2Client capi_v2_grpc.CarbonV2Client
+}
+
+// NewGrpc creates a new gRPC backend from the given configuration.
+func NewGrpc(cfg Config) (*GrpcBackend, error) {
+	b, err := New(cfg)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := grpc.Dial(cfg.Address, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, err
+	}
+	c := capi_v2_grpc.NewCarbonV2Client(conn)
+
+	return &GrpcBackend{
+		Backend:        b,
+		carbonV2Client: c,
+	}, nil
+}
+
+func makeMultiFetchRequestFromRenderRequest(request types.RenderRequest) *carbonapi_v2_pb.MultiFetchRequest {
+	frs := make([]*carbonapi_v2_pb.FetchRequest, 0, len(request.Targets))
+	for _, m := range request.Targets {
+		frs = append(frs, &carbonapi_v2_pb.FetchRequest{
+			Name:      m,
+			StartTime: request.From,
+			StopTime:  request.Until,
+		})
+	}
+	return &carbonapi_v2_pb.MultiFetchRequest{
+		Metrics: frs,
+	}
+}
+
+// Render fetches raw metrics from a backend.
+func (gb *GrpcBackend) Render(ctx context.Context, request types.RenderRequest) ([]types.Metric, error) {
+	t0 := time.Now()
+
+	multiFetchRequest := makeMultiFetchRequestFromRenderRequest(request)
+	request.Trace.AddMarshal(t0)
+
+	ctx, cancel := gb.setTimeout(ctx)
+	defer cancel()
+	stream, err := gb.carbonV2Client.Render(ctx, multiFetchRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	t1 := time.Now()
+	err = gb.enter(ctx)
+	request.Trace.AddLimiter(t1)
+	if err != nil {
+		return nil, err
+	}
+
+	var fetchedMetrics []types.Metric
+	for {
+		fetchResponse, err := stream.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		fetchedMetrics = append(fetchedMetrics, types.Metric{
+			Name:      fetchResponse.Name,
+			StartTime: fetchResponse.StartTime,
+			StopTime:  fetchResponse.StopTime,
+			StepTime:  fetchResponse.StepTime,
+			Values:    fetchResponse.Values,
+			IsAbsent:  fetchResponse.IsAbsent,
+		})
```

Comment:
Before streaming, what we did was to send the requests (which could have 1 to 10k metrics) to the backends, and merge them afterwards.
It is still the case as we do not aim to fix that in this specific PR.
I agree that it can be more optimized to merge the metrics on the go, but that needs some interface changes in the code and it can get a bit complex. Right now, It's not necessary to implement all this. It is something that we can consider after having streaming. WDYT?

Reply:
oh, so we still do split with streaming in this pr? I see, if it's gonna come as a separate change, then I guess yes. 

I was dreaming resolving the big query split with streaming. :P

Reply:
Yeah I believe we should do that in the end but I'm trying to keep the PRs smaller. Like for example gRPC is only implemented for Render ATM.

Generally, I think the following works after this should be:
* Implementing other rpcs as well and fully migrate to gRPC.
* Parallelize merging metrics with the Render stream for better performance and memory consumption
* Merging zipper and carbonapi for better performance and memory consumption 

As a result of these, we can remove the breaking logic.